### PR TITLE
Save nft on-chain

### LINKED ISSUE #23: Save NFT result on-chain
### When off-chain worker finishes a job, then a result has to be send on-chain

1. Send a signed transaction
2. Remove NFT from pending queue

### PR DESCRIPTION
**Pull request by korzewski**
Closes #23

### CODE REVIEW

File: pallets/nft/src/lib.rs [hunk 1]
```diff
@@ -107,22 +163,62 @@ pub mod pallet {
 			Ok(().into())
 		}
 
-		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 4))]
-		pub fn root_mint_nft(origin: OriginFor<T>, metadata: Cid, pending_nft: PendingNftOf<T>) -> DispatchResultWithPostInfo {
-			let _ = ensure_root(origin)?;
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 1))]
+		pub fn cancel_nft_request(origin: OriginFor<T>, pending_nft: PendingNftOf<T>, reason: TextMessage) -> DispatchResultWithPostInfo {
+			let _account_id = ensure_signed(origin)?;
```
Comment (aelesbao) [inline pallets/nft/src/lib.rs:168]: Is it possible to do the same validation without assigning?

File: pallets/nft/src/lib.rs [hunk 2]
```diff
@@ -107,22 +163,62 @@ pub mod pallet {
...
+
+			Self::remove_nft_from_pending_queue(pending_nft.clone())?;
+
+			Self::deposit_event(Event::CancelNftRequest(reason, pending_nft));
+			Ok(().into())
+		}
+
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(4, 5))]
+		pub fn mint_nft(origin: OriginFor<T>, metadata: TextMessage, pending_nft: PendingNftOf<T>) -> DispatchResultWithPostInfo {
+			let _account_id = ensure_signed(origin)?;
```
Comment (aelesbao) [inline pallets/nft/src/lib.rs:179]: Same here

**Review by aelesbao — APPROVED** [review]
Left a few comments. LGTM otherwise!

File: pallets/nft/src/lib.rs [hunk 3]
```diff
@@ -17,16 +27,48 @@ mod mock;
...
+	pub struct TestAuthId;
+
+	impl AppCrypto<MultiSigner, MultiSignature> for TestAuthId {
+		type RuntimeAppPublic = Public;
+		type GenericSignature = sp_core::sr25519::Signature;
+		type GenericPublic = sp_core::sr25519::Public;
+	}
+}
+
+pub type TextMessage = Vec<u8>;
```
[the remark this answers is not in the document]
Comment (korzewski) [reply pallets/nft/src/lib.rs:53]: I have realised that there were multiple "string" types, which are `Vec<u8>`. So decided to unify them as single type `TextMessage`. If you have a better name just let me know

File: pallets/nft/src/lib.rs (hunk 1 above)
Comment (korzewski) [reply pallets/nft/src/lib.rs:168]: Yeah, seems like - no warnings so far. Will remove this assignment

File: pallets/nft/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (aelesbao) [reply pallets/nft/src/lib.rs:53]: I see. I got a bit confused when I saw that the `dna` was a `TextMessage`. WDYT of using a more generic naming instead, like `ByteVector`?

**Review by aelesbao — APPROVED** [review]

**Pull request merged**